File: topKFrequent.py
'''
Given a non-empty list of words, return the k most frequent words. The output should be sorted from highest to lowest frequency, and if two words have the same frequency, the word with lower alphabetical order comes first. Input will contain only lower-case letters.

Example: 
Input: ["daily", "interview", "pro", "pro", 
"for", "daily", "pro", "problems"], k = 2
Output: ["pro", "daily"]
'''

import heapq, collections

class Solution:
    def topKFrequent(self, words, k):
        counts = collections.Counter(words)

        res = []
        heap = []
        for key, v in counts.items():
            heapq.heappush(heap, (-v, key))
        for _ in range(k, 0, -1):
            res.append(heapq.heappop(heap)[1])
        return res

# Solution: Heap & hashmap

class Solution:
    def topKFrequentE(self, words, k: int):
        hash = {}

        for word in words:
            if word in hash:
                hash[word] += 1
            else:
                hash[word] = 1

        heap = []
        for key in hash:
            heapq.heappush(heap, (-hash[key], key))
        
        res = []
        for i in range(k):
            popped = heapq.heappop(heap)
            res.append(popped[1])
        return res

# O(n) to create hashmap, O(n log n) to create the heap, and O(n) space.
words = ["daily", "interview", "pro", "pro", "for", "daily", "pro", "problems"]
k = 3
x = Solution().topKFrequentE(words, k)
print(x)

File: topKFrequent.js
/*
Given a non-empty list of words, return the k most frequent words. The output should be sorted from highest to lowest frequency, and if two words have the same frequency, the word with lower alphabetical order comes first. Input will contain only lower-case letters.

Example: 
Input: ["daily", "interview", "pro", "pro", 
"for", "daily", "pro", "problems"], k = 2
Output: ["pro", "daily"]
*/

// Solution 1: Map with reduce

var topKFrequent = function(words, k){
    const hash = words.reduce((map, word) => {
        if (map.has(word)) map.set(word, map.get(word) + 1)
        else map.set(word, 1)
        return map
    }, new Map())

    const sorted = [...hash].sort((a, b) => {
        if (a[1] > b[1]) return -1
        if (a[1] < b[1]) return 1
        if (a[0] > b[0]) return 1
        if (a[0] < b[0]) return -1
        return 1
    })
    return sorted.slice(0, k).map(([x]) => x)
}

// Solution 2: Hash & Sorting

var topKFrequent = (words, k) => {
    let hash = {}
    
    // insert uniq data inhashmap
    for (let i=0; i< words.length; i++){
        let temp = (hash[words[i]] ? hash[words[i]] : 0) + 1
        hash[words[i]] = temp
    }

    // loop through all the key of hashmap
    let result = Object.keys(hash).sort((next, previous) => {
        // First check for value count, if it is same check for key comparison
        if (hash[previous] - hash[next] == 0) {
            return next.localeCompare(previous)
        } else {
            return hash[previous] - hash[next]
        }
    })
    // Slice the index from 0 to Kth index which is length
    return result.slice(0, k)
}
